File: lu/lu_solver.c
#include <stdio.h>
#include <string.h>

#define NMAX 1000
#define abs(x) (((x) < 0) ? -(x) : (x))
#define eps 1e-16

int lucol(int n, double A[][NMAX], int p[]);
int sscol(int n, double A[][NMAX], int p[], double b[]);
int lurow(int n, double A[][NMAX], int p[]);
int ssrow(int n, double A[][NMAX], int p[], double b[]);

int leMatriz(double A[][NMAX], double b[NMAX]);
void swap(double *a, double *b);
void printMatrix(double A[][NMAX], int n);
void printVector(double v[], int n);

int main(int argc, char **argv) {
    int n = 0;
    int i;
    int p[NMAX];
    
    double b[NMAX];
    double A[NMAX][NMAX];
    
    n = leMatriz(A, b);
    if (argc > 1 && strcmp("-c", argv[1]) == 0) {
        printf("\nCalculando orientado a colunas");
        lucol(n, A, p);
        printf("\nLU:");
        printMatrix(A, n);
        sscol(n, A, p, b);
        printf("\nx:");
        printVector(b, n);
    }
    else {
        printf("\nCalculando orientado a linhas");
        if (!lurow(n, A, p)) {
            printf("\nLU:");
            printMatrix(A, n);
        }
        else {
            printf("\n\nDeu pau\n");
            return -1;
        }
        if (!ssrow(n, A, p, b)) {
            printf("\nx:");
            printVector(b, n);
        }
        else {
            printf("\n\nDeu pau\n");
            return -1;
        }
    }
    

    return 0;
}

int leMatriz(double A[][NMAX], double b[NMAX]) {
    int n, k, i, j, nsquare;
    scanf("%d", &n);
    nsquare = n * n;
    for (k = 0; k < nsquare; k++) {
        scanf("%d %d", &i, &j);
        scanf("%lf", &A[i][j]);
    }
    for (k = 0; k < n; k++) {
        scanf("%d", &i);
        scanf("%lf", &b[i]);
    }
    return n;
}

int lurow(int n, double A[][NMAX], int p[]) {
    int k, i, j, imax;

    for (k = 0; k < n; k++) {
        imax = k;
        for (i = k + 1; i < n; i++) 
            if (abs(A[i][k]) > abs(A[imax][k])) 
                imax = i;
        if (abs(A[imax][k]) < eps) 
            return -1;
        if (imax != k)
            for (j = 0; j < n; j++)
                swap(&A[imax][j], &A[k][j]);
        p[k] = imax;
        for (i = k + 1; i < n; i++) {
            A[i][k] = A[i][k] / A[k][k];
            for (j = k + 1; j < n; j++)
                A[i][j] = A[i][j] - A[k][j] * A[i][k];
        }
    }     
    return 0;
}

int ssrow(int n, double A[][NMAX], int p[], double b[]) {
    int i, k;

    for (i = 0; i < n; i++) 
        if (p[i] != i)
            swap(&b[p[i]], &b[i]);

    for (i = 0; i < n; i++)
        for (k = 0; k < i; k++)
            b[i] = b[i] - b[k] * A[i][k];

    for (i = n - 1; i >= 0; i--) {
        for (k = n - 1; k > i; k--)
            b[i] = b[i] - b[k] * A[i][k];
        if (abs(A[i][i]) < eps)
            return -1;
        b[i] = b[i] / A[i][i];
    }

    return 0;
}

int lucol(int n, double A[][NMAX], int p[]) {
    int k, i, j, imax;

    for (k = 0; k < n; k++) {
        imax = k;
        for (i = k + 1; i < n; i++) 
            if (abs(A[i][k]) > abs(A[imax][k])) 
                imax = i;

        if (abs(A[imax][k]) < eps) 
            return -1; 
        if (imax != k)
            for (j = 0; j < n; j++)
                swap(&A[imax][j], &A[k][j]);
        p[k] = imax;


        for (i = k + 1; i < n; i++)
            A[i][k] = A[i][k] / A[k][k];

        for (j = k + 1; j < n; j++)            
            for (i = k + 1; i < n; i++)
                A[i][j] = A[i][j] - A[k][j] * A[i][k];
    }     
    return 0;
}

int sscol(int n, double A[][NMAX], int p[], double b[]) {
    int i, k;

    for (i = 0; i < n; i++)
        if (p[i] != i)
            swap(&b[p[i]], &b[i]);

    for (k = 0; k < n; k++)
        for (i = k + 1; i < n; i++)
            b[i] = b[i] - b[k] * A[i][k];

    for (k = n - 1; k >= 0; k--) {
        if (abs(A[k][k]) < eps)
            return -1;
        b[k] = b[k] / A[k][k];
        for (i = k - 1; i >= 0; i--)
            b[i] = b[i] - b[k] * A[i][k];
    }

    return 0;
}

void swap(double *a, double *b) {
    double temp = *a;
    *a = *b;
    *b = temp;
}

void printMatrix(double A[][NMAX], int n) {
    int i, j;
    printf("\n");
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++)
            printf("% 3.2f ", A[i][j]);
        printf("\n");
    }
}

void printVector(double v[], int n) {
    int i;
    printf("\n");
    for (i = 0; i < n; i++)
        printf("% 3.2f\n", v[i]);
}

File: lu/fortran/lu_solver.f90
module constants
    double precision, parameter :: eps = epsilon(0d0)
    integer, parameter :: NMAX = 1000
end module

program lu_solver
    use constants

    implicit none
    integer, external :: lurow, ssrow, lucol, sscol
    integer :: res, i, j

    double precision, dimension(NMAX, NMAX) :: A
    double precision, dimension(NMAX) :: b
    integer, dimension(NMAX) :: p 
    integer :: n = 3
    integer :: lda = 1

    call readMatrix(A, b, n)   
 
    print *, "A before:"
    call printMatrix(A, n)
    print *,"p before:"
    call printiVector(p, n)
    print *,"b before:"
    call printdVector(b, n)

    res = lucol(n, lda, A, p)

    print "(/,a)", "A after:"
    call printMatrix(A, n)
    print *, "p after:"
    call printiVector(p, n)
    print *, "res: ", res

    res = sscol(n, lda, A, p, b)

    print "(/,a)", "A final:"
    call printMatrix(A, n)
    print *, "b final:"
    call printdVector(b, n)
    print *, "res: ", res
    
end program

integer function lurow(n, lda, A, p)
    use constants
    integer, intent(in) :: n, lda
    integer, intent(inout) :: p(NMAX)
    double precision, intent(inout) :: A(lda:NMAX, lda:NMAX)

    integer :: k, i, j, imax

    do k = 1, n
        imax = k
        do i = k + 1, n
            if (abs(A(i, k)) > abs(A(imax, k))) imax = i
        end do

        if (abs(A(imax,k)) < eps) then
            lurow = -1 
            return
        end if
    
        if (imax /= k) then
            do j = 1, n
                call swap(A(imax, j), A(k, j))
            end do
        end if
        p(k) = imax

        do i = k + 1, n
            A(i, k) = A(i,k) / A(k, k)
            do j = k + 1, n
                A(i, j) = A(i, j) - A(k, j) * A(i, k)
            end do
        end do
    end do
    lurow = 0
end function

integer function ssrow(n, lda, A, p, b)
    use constants
    implicit none
    integer, intent(in) :: n, lda
    integer, intent(inout) :: p(NMAX)
    double precision, intent(inout) :: A(lda:NMAX, lda:NMAX)
    double precision, intent(inout) :: b(lda:NMAX)

    integer :: k, i

    do i = 1, n
        if (p(i) /= i) call swap(b(p(i)), b(i))
    end do

    do i = 1, n
        do k = 1, i - 1
            b(i) = b(i) - b(k) * A(i, k)
        end do
    end do

    do i = n, 1, -1
        do k = n, i + 1, -1
            b(i) = b(i) - b(k) * A(i, k)
        end do
        if (abs(A(i,i)) < eps) then
            ssrow = -1
            return
        end if
        b(i) = b(i) / A(i, i)
    end do
    ssrow = 0
end function

integer function lucol(n, lda, A, p)
    use constants
    implicit none
    integer, intent(inout) :: n, lda
    integer, intent(inout) :: p(NMAX)
    double precision, intent(inout) :: A(lda:NMAX, lda:NMAX)

    integer :: k, i, j, imax

    do k = 1, n
        imax = k
        do i = k + 1, n
            if (abs(A(i, k)) > abs(A(imax, k))) imax = i
        end do
        
        if (abs(A(imax, k)) < eps) then
            lucol = -1
            return
        end if

        if (imax /= k) then
            do j = 1, n
                call swap(A(imax, j), A(k, j))
            end do
        end if
        p(k) = imax

        do i = k + 1, n
            A(i, k) = A(i, k) / A(k, k)
        end do

        do j = k + 1, n
            do i = k + 1, n
                A(i, j) = A(i, j) - A(k, j) * A(i, k)
            end do
        end do
    end do
    lucol = 0
end function

integer function sscol(n, lda, A, p, b)
    use constants
    implicit none
    integer, intent(in) :: n, lda
    integer, intent(inout) :: p(NMAX)
    double precision, intent(inout) :: A(lda:NMAX, lda:NMAX)
    double precision, intent(inout) :: b(lda:NMAX)

    integer :: k, i

    do i = 1, n
        if (p(i) /= i) call swap(b(p(i)), b(i))
    end do

    do k = 1, n
        do i = k + 1, n
            b(i) = b(i) - b(k) * A(i, k)
        end do
    end do
    
    do k = n, 1, -1
        if (abs(A(k, k)) < eps) then
            sscol = -1
            return
        end if
        b(k) = b(k) / A(k, k)
        do i = k - 1, 1, -1
            b(i) = b(i) - b(k) * A(i, k)
        end do
    end do
    sscol = 0
end function


subroutine printMatrix(A, n)
    use constants
    implicit none
    
    integer, intent(in) :: n
    double precision, intent(in) :: A(NMAX, NMAX)

    integer :: i, j

    do i = 1, n
        do j = 1, n
            print "(f7.3$)", A(i, j)
        end do
        print *,""
    end do
end subroutine

subroutine printdVector(v, n)
    use constants
    implicit none
    integer, intent(in) :: n
    double precision, intent(in) :: v(NMAX)
    integer :: i
    do i = 1, n
        print "(f7.3$)", v(i)
    end do
end subroutine

subroutine printiVector(v, n)
    use constants
    implicit none
    integer, intent(in) :: n
    integer, intent(in) :: v(NMAX)
    integer :: i
    do i = 1, n
        print "(i3)", v(i)
    end do
end subroutine


subroutine swap(a, b)
    implicit none
    double precision, intent(inout) :: a, b
    double precision :: temp

    temp = a
    a = b
    b = temp
end subroutine

subroutine readMatrix(A, b, n)
    use constants
    implicit none
    double precision, dimension(NMAX, NMAX), intent(out) :: A
    double precision, dimension(NMAX), intent(out) :: b
    integer, intent(out) :: n
    integer :: i, j, k
    integer :: nsquare
    read *, n
    nsquare = n ** 2
    do k = 1, nsquare
        read *, i, j, A(i, j)
    end do
    do k = 1, n
        read *, i, b(i)
    end do
end subroutine
